Sources/CacheAdvance/CacheAdvance.swift

```diff
@@ -0,0 +1,215 @@
+    ///   - maximumBytes: The maximum size of the cache, in bytes. Logs larger than this size will fail to append to the store.
+    ///   - shouldOverwriteOldMessages: When `true`, once the on-disk store exceeds maximumBytes, new entries will replace the oldest entry.
+    ///
+    /// - Warning: `maximumBytes` must be consistent for the life of a cache. Changing this value after logs have been persisted to a cache leads to undefined behavior.
+    public init(
+        file: URL,
+        maximumBytes: Bytes,
+        shouldOverwriteOldMessages: Bool)
+        throws
+    {
+        self.file = file
+        self.maximumBytes = maximumBytes
+        self.shouldOverwriteOldMessages = shouldOverwriteOldMessages
+
+        writer = try FileHandle(forWritingTo: file)
+        reader = try FileHandle(forReadingFrom: file)
+
+        lengthOfMessageSuffix = {
+            if shouldOverwriteOldMessages {
+                /// The message suffix requires space for both `endOfNewestMessageMarker` and `offsetOfFirstMessage` after each message when rolling is enabled.
+                return Bytes(Data.endOfNewestMessageMarker.count) + Bytes(Data.bytesStorageLength)
+            } else {
+                /// The message suffix requires space for `endOfNewestMessageMarker` after each message.
+                return Bytes(Data.endOfNewestMessageMarker.count)
+            }
+        }()
+    }
+
+    deinit {
+        try? writer.close()
+        try? reader.close()
+    }
+
+    // MARK: Public
+
+    public let file: URL
+
+    /// Appends a message to the cache.
+    /// - Parameter message: A message to write to disk.
+    public func append(message: T) throws {
+        try setUpFileHandlesIfNecessary()
+
+        let encodedMessage = EncodableMessage(message: message, encoder: encoder)
+        let messageData = try encodedMessage.encodedData()
+        let messageLength = Bytes(messageData.count)
+
+        guard messageLength > 0 else {
+            // The message length has the same value as as our `endOfNewestMessageMarker`.
+            // Storing this message could cause data corruption by fooling the cache into thinking the prior message is the last in the cache.
+            // JSON values (even empty ones) are always encoded with length > 0. If this condition is hit, the message is clearly corrupt.
+            throw CacheAdvanceWriteError.messageCorrupted
+        }
+
+        let bytesNeededToStoreMessage = messageLength + lengthOfMessageSuffix
+        guard bytesNeededToStoreMessage <= maximumBytes && bytesNeededToStoreMessage < Bytes(MessageSpan.max) else {
+            // The message is too long to be written to a cache of this size.
+            throw CacheAdvanceWriteError.messageDataTooLarge
+        }
+
+        let cacheHasSpaceForNewMessageBeforeEndOfFile = writer.offsetInFile + bytesNeededToStoreMessage <= maximumBytes
```

Comment:
I can write a file that **is** `maximumBytes` long.

If I make this a `<` rather than a `<=`, my unit tests that try to write a message to a file that is the same length as `maximumBytes` fails.

I'm reasonably confident I don't have an off-by-one error here... unless I've managed to make the same off-by-one error in both my tests and the production code. 0-indexing is hard 😅.